### PR TITLE
DBZ-6399 Fix Date and Time value resolution w/time zones

### PR DESCRIPTION
**Pull request by Naros**
https://issues.redhat.com/browse/DBZ-6399

### Timezone influencers
To efficiently test various configurations of time zones in the future, I took the liberty to introduce two system variables called `source.time_zone` and `sink.time_zone`.  By default, the source databases are set to `UTC` while the sink databases are started in `PST` or `America/Los_Angeles`. This means that all temporal-based tests operate across timezone boundaries to try and catch potential time-zone-related bugs faster moving forward.

### Type adjustments
All types related to `Date` and `Time` have been adjusted to leverage the `database.time_zone` configuration option correctly, which guarantees that the value written to the sink database will not be negatively affected by time zones.  To accomplish this, a new `AbstractTemporalType` was introduced in conjunction with the `Type#configure` method to make the configured timezone available to the type subsystem.

### CODE REVIEW

**Review by jpechane — APPROVED** [review]
@Naros LGTM! I'll merge it tomorrow morning if no feedback from reporter is provided.

**Comment by jpechane** [comment]
@Naros It seems that the CI timing issue is the problem. I modified the startup script to introduce a delay to let the job to start up. Now it is set to 15 secs. We can try even lower value. If we'd like to have it done best then it is necessary to write a T-SQL code (loop) that will check the status of the job and wait till it is running.

File: src/main/java/io/debezium/connector/jdbc/dialect/DatabaseDialect.java [hunk 1]
```diff
@@ -143,6 +144,9 @@ public interface DatabaseDialect {
      */
     int getMaxVarbinaryLength();
 
+    boolean isConnectionTimeZoneSet();
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/dialect/DatabaseDialect.java:147]: Please add Javadoc to the new ethods.

File: src/main/java/io/debezium/connector/jdbc/type/connect/ConnectTimeType.java [hunk 2]
```diff
@@ -58,15 +49,23 @@ public void bind(Query<?> query, int index, Schema schema, Object value) {
             query.setParameter(index, null);
         }
         else if (value instanceof Date) {
-            final ZonedDateTime zdt = toZonedDateTime((Date) value);
-            query.setParameter(index, zdt, StandardBasicTypes.ZONED_DATE_TIME_WITHOUT_TIMEZONE);
+            final LocalTime localTime = DateTimeUtils.toLocalTimeFromUtcDate((Date) value);
+            final LocalDateTime localDateTime = localTime.atDate(LocalDate.now());
+            if (getDialect().isConnectionTimeZoneSet() || getDialect().isJdbcTimeZoneSet()) {
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/type/connect/ConnectTimeType.java:54]: Should not there be a default method like `isTimezoneAvailable()` or make `getDatabaseTimeZone` to return `Optional`?

File: src/main/java/io/debezium/connector/jdbc/type/debezium/AbstractDebeziumTimeType.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            final LocalTime localTime = getLocalTime((Number) value);
+            final LocalDateTime localDateTime = localTime.atDate(LocalDate.now());
+            if (getDialect().isConnectionTimeZoneSet() || getDialect().isJdbcTimeZoneSet()) {
+                query.setParameter(index, localDateTime.atZone(getDatabaseTimeZone().toZoneId()));
+            }
+            else {
+                query.setParameter(index, localDateTime);
+            }
+            // TODO: When not setting Hibernate JDBC_TIME_ZONE, this works
+//            final LocalTime localTime = getLocalTime((Number) value);
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/type/debezium/AbstractDebeziumTimeType.java:54]: I am not sure how to interret the comment, is it something to be fixed in future or removed?

File: src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java [hunk 4]
```diff
@@ -65,15 +68,33 @@ public void bind(Query<?> query, int index, Schema schema, Object value) {
...
-            query.setParameter(index, offsetTime.atDate(LocalDate.now()).toInstant());
+            final ZonedDateTime zdt = OffsetTime.parse((String) value, ZonedTime.FORMATTER).atDate(LocalDate.now()).toZonedDateTime();
+            if (getDialect().isJdbcTimeZoneSet()) {
+                query.setParameter(index, zdt, StandardBasicTypes.ZONED_DATE_TIME_WITH_TIMEZONE);
+            }
+            else {
+                if (getDialect().isConnectionTimeZoneSet()) {
+                    query.setParameter(index, zdt.withZoneSameInstant(getDatabaseTimeZone().toZoneId()));
+                }
+                else if (getDialect() instanceof SqlServerDatabaseDialect) {
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java:79]: I think theis should be extracted into dialect itself if necessary.

File: src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java [hunk 5]
```diff
@@ -65,15 +68,33 @@ public void bind(Query<?> query, int index, Schema schema, Object value) {
...
+                }
+            }
         }
         else {
             throwUnexpectedValue(value);
         }
     }
 
+    private int getJdbcType(DatabaseDialect dialect) {
+        if (dialect instanceof SqlServerDatabaseDialect) {
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java:93]: Same as above

File: src/main/java/io/debezium/connector/jdbc/type/connect/ConnectTimeType.java (hunk 2 above)
Comment (Naros) [reply src/main/java/io/debezium/connector/jdbc/type/connect/ConnectTimeType.java:54]: That was my original intent, but the behavior in `ZonedTimeType` differs based on which of these values are set individually, so I didn't want to necessarily introduce any confusion by doing this.  wdyt?

File: src/main/java/io/debezium/connector/jdbc/type/debezium/AbstractDebeziumTimeType.java (hunk 3 above)
Comment (Naros) [reply src/main/java/io/debezium/connector/jdbc/type/debezium/AbstractDebeziumTimeType.java:54]: I've removed this, this was missed in PR polish.

File: src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java (hunk 4 above)
Comment (Naros) [reply src/main/java/io/debezium/connector/jdbc/type/debezium/ZonedTimeType.java:79]: It is moved into a SQL Server Type implementation that extends and overrides the SQL Server-specific bits.

**Comment by jpechane** [comment]
@Naros Applied, thanks. Huge piece of work!

**Pull request merged**